File: packages/react-native-autoplay/android/src/main/java/com/margelo/nitro/swe/iternio/reactnativeautoplay/HybridPointOfInterestTemplate.kt
package com.margelo.nitro.swe.iternio.reactnativeautoplay

import com.margelo.nitro.core.Promise
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.AndroidAutoTemplate
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.PointOfInterestTemplate

class HybridPointOfInterestTemplate : HybridPointOfInterestTemplateSpec() {

    override fun createPointOfInterestTemplate(config: PointOfInterestTemplateConfig) {
        val context = AndroidAutoSession.getRootContext()
            ?: throw IllegalArgumentException("createPointOfInterestTemplate failed, carContext not found")

        val template = PointOfInterestTemplate(context, config)
        AndroidAutoTemplate.setTemplate(config.id, template)
    }

    override fun updatePointOfInterestTemplateItems(
        templateId: String, items: Array<PointOfInterest>
    ): Promise<Unit> {
        return Promise.async {
            val template = AndroidAutoTemplate.getTemplate<PointOfInterestTemplate>(templateId)
            template.updateItems(items)
        }
    }
}

File: packages/react-native-autoplay/ios/hybrid/HybridPointOfInterestTemplate.swift
//
//  HybridPointOfInterestTemplate.swift
//  Pods
//

import NitroModules

class HybridPointOfInterestTemplate: HybridPointOfInterestTemplateSpec {
    func createPointOfInterestTemplate(config: PointOfInterestTemplateConfig) throws {
        let template = PointOfInterestTemplate(config: config)

        try RootModule.withTemplateStore { templateStore in
            templateStore.addTemplate(
                template: template,
                templateId: config.id
            )
        }
    }

    func updatePointOfInterestTemplateItems(
        templateId: String,
        items: [PointOfInterest]
    ) throws -> Promise<Void> {
        return Promise.async {
            try await MainActor.run {
                try RootModule.withAutoPlayTemplate(templateId: templateId) {
                    (template: PointOfInterestTemplate) in
                    template.updateItems(items: items)
                }
            }
        }
    }
}
